File: Java/laborator3JavaGUI/src/main/java/service/Service.java
package service;

import domain.Cursa;
import domain.Oficiu;
import domain.Participant;
import domain.ParticipantCursa;
import domain.validator.ParticipantCursaValidator;
import domain.validator.Validator;
import domain.validator.ValidatorException;
import javafx.util.Pair;
import repo.*;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Service {
    private OficiuRepository oficiuRepository;
    private ParticipantRepository participantRepository;
    private CursaRepository cursaRepository;
    private ParticipantCursaRepository participantCursaRepository;

    public Service(OficiuRepository oficiuRepository, ParticipantRepository participantRepository, CursaRepository cursaRepository, ParticipantCursaRepository participantCursaRepository) {
        this.oficiuRepository = oficiuRepository;
        this.participantRepository = participantRepository;
        this.cursaRepository = cursaRepository;
        this.participantCursaRepository = participantCursaRepository;
    }

    //Oficiu
    public void saveOficiu(Oficiu oficiu){
        oficiuRepository.save(oficiu);
    }

    public void deleteOficiu(String id){
        oficiuRepository.delete(id);
    }

    public void updateOficiu(String id, Oficiu oficiu){
        oficiuRepository.update(id, oficiu);
    }

    public Oficiu findOficiu(String id){
        return oficiuRepository.findOne(id);
    }

    public Oficiu findOficiuByUserPass(String username, String password){
        return oficiuRepository.findByUserPass(username, password);
    }

    public List<Oficiu> findAllOficiu(){
        List<Oficiu> all = new ArrayList<>();
        oficiuRepository.findAll().forEach(all::add);
        return all;
    }

    //Cursa
    public void saveCursa(Cursa cursa){
        cursaRepository.save(cursa);
    }

    public void deleteCursa(String id){
        cursaRepository.delete(id);
    }

    public void updateCursa(String id, Cursa cursa){
        cursaRepository.update(id, cursa);
    }

    public Cursa findCursa(String id){
        return cursaRepository.findOne(id);
    }

    public List<Cursa> findAllCursa(){
        List<Cursa> all = new ArrayList<>();
        cursaRepository.findAll().forEach(all::add);
        return all;
    }

    public List<Cursa> findAllCursa(Integer capMotor){
        List<Cursa> all = new ArrayList<>();
        cursaRepository.findAll(capMotor).forEach(all::add);
        return all;
    }

    //Participant
    public void saveParticipant(Participant participant){
        participantRepository.save(participant);
    }

    public void deleteParticipant(String id){
        participantRepository.delete(id);
    }

    public void updateParticipant(String id, Participant participant){
        participantRepository.update(id, participant);
    }

    public Participant findParticipant(String id){
        return participantRepository.findOne(id);
    }

    public Participant findParticipant(String nume, String echipa, Integer capMotor){
        return participantRepository.findOneByNumeEchipaCapMotor(nume, echipa, capMotor);
    }

    public List<Participant> findAllParticipant(){
        List<Participant> all = new ArrayList<>();
        participantRepository.findAll().forEach(all::add);
        return all;
    }

    public List<Participant> findAllParticipant(String echipa){
        List<Participant> all = new ArrayList<>();
        for(Participant participant : participantRepository.findAll()){
            if(participant.getEchipa().equals(echipa))
                all.add(participant);
        }
        return all;
    }

    public String findAllEchipe(){
        String echipe = "Echipele participantilor:\n";
        List<String> list = new ArrayList<>();
        for(Participant participant : findAllParticipant()) list.add(participant.getEchipa());
        List<String> result = list.stream().distinct().collect(Collectors.toList());
        for(String el : result){
            echipe = echipe + el + "\n";
        }
        return echipe;
    }

    //ParticipantCursa
    public void saveParticipantCursa(ParticipantCursa participantCursa){
        participantCursaRepository.save(participantCursa);
    }

    public void deleteParticipantCursa(String idParticipant, String idCursa){
        participantCursaRepository.delete(new Pair<>(idParticipant, idCursa));
    }

    public void updateParticipantCursa(String idParticipant, String idCursa, ParticipantCursa participantCursa){
        participantCursaRepository.update(new Pair<>(idParticipant, idCursa), participantCursa);
    }

    public ParticipantCursa findParticipantCursa(String idParticipant, String idCursa){
        return participantCursaRepository.findOne(new Pair<>(idParticipant, idCursa));
    }

    public List<ParticipantCursa> findAllParticipantCursa(){
        List<ParticipantCursa> all = new ArrayList<>();
        participantCursaRepository.findAll().forEach(all::add);
        return all;
    }


}

File: C#/laborator3CSharpGUI/lab3/lab3/service/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using lab3.domain;
using lab3.repo;

namespace lab3.service
{
    class Service
    {
        private OficiuRepository OficiuRepository;
        private ParticipantRepository ParticipantRepository;
        private CursaRepository CursaRepository;
        private ParticipantCursaRepository ParticipantCursaRepository;

        public Service(OficiuRepository oficiuRepository, ParticipantRepository participantRepository, CursaRepository cursaRepository, ParticipantCursaRepository participantCursaRepository)
        {
            OficiuRepository = oficiuRepository;
            ParticipantRepository = participantRepository;
            CursaRepository = cursaRepository;
            ParticipantCursaRepository = participantCursaRepository;
        }

        //Oficiu
        public Oficiu findOficiu(string id)
        {
            return OficiuRepository.findOne(id);
        }

        public Oficiu findOficiu(string username, string password)
        {
            return OficiuRepository.findOne(username, password);
        }

        public IList<Oficiu> findAllOficiu()
        {
            IList<Oficiu> oficiuList = new List<Oficiu>();
            foreach (Oficiu oficiu in OficiuRepository.findAll())
            {
                oficiuList.Add(oficiu);
            }
            return oficiuList;
        }

        //Participant
        public Participant findParticipant(string id)
        {
            return ParticipantRepository.findOne(id);
        }

        public Participant findParticipant(string nume, string echipa, int capMotor)
        {
            return ParticipantRepository.findOne(nume, echipa, capMotor);
        }

        public IList<Participant> findAllParticipant()
        {
            IList<Participant> participantList = new List<Participant>();
            foreach (Participant participant in ParticipantRepository.findAll())
            {
                participantList.Add(participant);
            }
            return participantList;
        }

        public IList<Participant> findAllParticipant(string echipa)
        {
            IList<Participant> participantList = new List<Participant>();
            foreach (Participant participant in ParticipantRepository.findAll(echipa))
            {
                participantList.Add(participant);
            }
            return participantList;
        }

        //Cursa
        public Cursa findCursa(string id)
        {
            return CursaRepository.findOne(id);
        }

        public IList<Cursa> findAllCursa()
        {
            IList<Cursa> cursaList = new List<Cursa>();
            foreach (Cursa cursa in CursaRepository.findAll())
            {
                cursaList.Add(cursa);
            }
            return cursaList;
        }

        public IList<Cursa> findAllCursa(int capMotor)
        {
            IList<Cursa> cursaList = new List<Cursa>();
            foreach (Cursa cursa in CursaRepository.findAll(capMotor))
            {
                cursaList.Add(cursa);
            }
            return cursaList;
        }

        //ParticipantCursa
        public void saveParticipantCursa(ParticipantCursa participantCursa)
        {
            ParticipantCursaRepository.save(participantCursa);
        }
    }
}
